### ISSUE TITLE
explicit_init - creating an instance from a meta-type

### ISSUE THREAD

**Issue by evgeniyd** [realm/SwiftLint#984]
Creating an instance from meta-types via explicit `init()` is a common idiom. Currently, swiftlint highlights instructions like this when `explicit_init` rule is opted-in.

    struct SomeStruct {
        var a: Int = 0
    }
    
    class SomeClass {
        var a: Int = 0
        
        required init() {
            a = 1
        }
    }
    
    let metaType = SomeClass.self // SomeClass.Type
    let instance = metaType.init()
    
    let structMetaType =  SomeStruct.self // SomeStruct.Type
    let structInstance = structMetaType.init()

**Comment by marcelofabri** [comment]
Are you sure this example triggers? I couldn't reproduce on SwiftLint 0.13.2:

```
$ swiftlint lint --path file.swift          
Loading configuration from '.swiftlint.yml'
Linting Swift files at path file.swift
Linting 'file.swift' (1/1)
file.swift:2:2: error: Variable Name Violation: Variable name should be between 3 and 40 characters long: 'a' (variable_name)
file.swift:6:2: error: Variable Name Violation: Variable name should be between 3 and 40 characters long: 'a' (variable_name)
Done linting! Found 2 violations, 2 serious in 1 file.
```

**Comment by evgeniyd** [comment]
@marcelofabri my fault, I didn't do linting, just shown the idiom. 

Here is the snipped that does produce the warning:

    import UIKit.UIView
    
    public enum KeyImage {
        case standard
        case customView(type: UIView.Type)
    }
    
    extension KeyImage {
        public func view() -> UIView {
            switch self {
            case .standard:
                return StandardView()
            case .customView(let ViewType):
                return ViewType.init()
            }
        }
    }
    
    class StandardView: UIView { }
    class MyView: UIView { }
    
    KeyImage.standard.view()
    KeyImage.customView(type: MyView.self).view()

Output:

```
$ swiftlint lint --path file.swift
Loading configuration from '.swiftlint.yml'
Linting Swift files at path file.swift
Linting 'file.swift' (1/1)
file.swift:14:28: warning: Explicit Init Violation: Explicitly calling .init() should be avoided. (explicit_init)
Done linting! Found 1 violation, 0 serious in 1 file.
```

**Comment by marcelofabri** [comment]
The rule currently uses a regex to check if it's a type or metatype, so changing `ViewType` to `viewType` fixes the issue.

**Comment by evgeniyd** [comment]
@marcelofabri Thank you for explanation. I thought it is possible to infer the type, as swiftlint hooked into Clang and SourceKit. 

Unfortunately, lowercasing the var name here loses the semantics I'd like to preserve. Nevertheless, this particular example, could be fixed in this way, I have other cases, where I was in need to go with `// swiftlint:disable:this explicit_init`.

**Comment by jpsim** [comment]
> I thought it is possible to infer the type, as swiftlint hooked into Clang and SourceKit.

Possible yes, but not implemented. Would require a new "compiled" mode that SwiftLint could run in such as the one described in https://github.com/realm/SwiftLint/pull/829#issuecomment-263008789.
